[FLINK-21189][runtime] Added support for concurrently caught exceptions to exception history

XComp opened the pull request:
## What is the purpose of the change

The goal is to not only provide the root cause of a task or global failure but exceptions that were caught concurrently in other `Executions` which would be affected by the partial or full restart of the job.

## Brief change log

* Introduced `RooExceptionHistoryEntry`
* Introduced `ExceptionHistoryEntryExtractor` that deals with creating the history entries
* Created exceptionhistory sub-package

## Verifying this change

This change added tests and can be verified as follows:

* `ExceptionHistoryEntryExtractorTest` was added dealing with the actual entry extraction

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? REST API docs

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/ExceptionHistoryService.java:
```diff
@@ -0,0 +1,103 @@
+import java.util.List;
+
+/** {@code ExceptionHistoryService} collects failure causes. */
+public class ExceptionHistoryService {
```
Introducing an extra component for this is fine, but the initial introduction should be part of a separate ticket.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/DefaultExecutionGraph.java:
```diff
@@ -377,6 +378,28 @@ public ComponentMainThreadExecutor getJobMasterMainThreadExecutor() {
 
+    @Override
+    public ExecutionVertex getExecutionVertex(final ExecutionVertexID executionVertexId) {
+        final Map<JobVertexID, ExecutionJobVertex> executionJobVertices = getAllVertices();
+        Preconditions.checkArgument(
+                executionJobVertices.containsKey(executionVertexId.getJobVertexId()),
+                "The passed executionVertexId {} is invalid: There is no ExecutionJobVertex with the associated JobVertexId {}.",
+                executionVertexId,
+                executionVertexId.getJobVertexId());
+
+        final ExecutionJobVertex executionJobVertex =
+                executionJobVertices.get(executionVertexId.getJobVertexId());
```
Do we really need to add another method to the EG for something that is already well supported with existing APIs?
```
AccessExecutionVertex vertex = accessExecutionGraph.getJobVertex(executionVertexId.getJobVertexId())[executionVertexId.getSubtaskIndex()];
```

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertexProvider.java:
```diff
@@ -0,0 +1,47 @@
+     * @throws IllegalArgumentException if the passed {@code ExecutionVertexID} is not associated to
+     *     a {@code ExecutionVertex} handled by this provider.
+     */
+    T getExecutionVertex(ExecutionVertexID executionVertexId);
```
This interface seems weird given that the history service only cares about executions.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/ExceptionHistoryEntryFactory.java:
```diff
@@ -0,0 +1,30 @@
+public interface ExceptionHistoryEntryFactory {
+    Optional<ExceptionHistoryEntry> create(ExecutionVertexID executionVertexID);
+
+    Iterable<ExceptionHistoryEntry> extractAllFailures();
```
This is unused in production code. It also seems like quite an odd interface TBH, and would prefer the EG to actively pass in all failure causes.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/ExceptionHistoryService.java:
```diff
@@ -0,0 +1,103 @@
+        log.debug("Archive global failure.", throwable);
+    }
+
+    public void registerLocalFailure(FailureHandlingResult failureHandlingResult) {
```
I'm not convinced that passing in an AccessExecution makes testing much more difficult. And then we could forego this whole execption entry factory business.

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/exceptionhistory/ExceptionHistoryEntryExtractorTest.java:
```diff
@@ -0,0 +1,159 @@
+import org.apache.flink.runtime.taskmanager.TaskExecutionState;
+import org.apache.flink.util.TestLogger;
+
+import org.apache.flink.shaded.curator4.com.google.common.collect.Iterables;
```
wrong import

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/exceptionhistory/ExceptionHistoryEntryExtractorTest.java:
```diff
@@ -0,0 +1,159 @@
+        final Range<Long> rootFailureTimeframe = triggerFailure(rootExecutionVertex, rootException);
+
+        final Throwable otherException = new IllegalStateException("Expected other failure");
+        final ExecutionVertex otherExecutionVertex = extractExecutionVertex(1);
```
```suggestion
        final ExecutionVertex concurrentlyFailedExecutionVertex = extractExecutionVertex(1);
```

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/exceptionhistory/ExceptionHistoryEntryExtractorTest.java:
```diff
@@ -0,0 +1,159 @@
+    }
+
+    private Range<Long> triggerFailure(ExecutionVertex executionVertex, Throwable throwable) {
+        final long startInMillis = System.currentTimeMillis();
```
shouldn't be necessary; the failure timestamp is the FAILED state timestamp of the vertex

XComp replied:
Good point 👍

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/exceptionhistory/ExceptionHistoryEntryExtractorTest.java:
```diff
@@ -0,0 +1,159 @@
+
+    @Test
+    public void extractGlobalFailure() {
+        final Throwable otherException0 = new RuntimeException("Expected other failure #0");
```
```suggestion
        final Throwable concurrentException0 = new RuntimeException("Expected other failure #0");
```

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/exceptionhistory/ExceptionHistoryEntryExtractorTest.java:
```diff
@@ -0,0 +1,159 @@
+        final Range<Long> otherFailureTimeframe0 =
+                triggerFailure(otherExecutionVertex0, otherException0);
+
+        final Throwable otherException1 = new IllegalStateException("Expected other failure");
```
```suggestion
        final Throwable concurrentException1 = new IllegalStateException("Expected other failure");
```

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/ExceptionHistoryEntryTest.java:
```diff
@@ -24,6 +24,8 @@
 import org.apache.flink.runtime.executiongraph.ErrorInfo;
 import org.apache.flink.runtime.executiongraph.ExecutionAttemptID;
 import org.apache.flink.runtime.executiongraph.IOMetrics;
+import org.apache.flink.runtime.scheduler.exceptionhistory.ExceptionHistoryEntry;
```
this test should also be moved

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/ExceptionHistoryEntryExtractor.java:
```diff
@@ -0,0 +1,105 @@
+
+    public RootExceptionHistoryEntry extractLocalFailure(
+            Map<JobVertexID, ExecutionJobVertex> executionJobVertices,
+            ExecutionVertexID executionVertexId,
```
```suggestion
            ExecutionVertexID failedExecutionVertexId,
```

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/ExceptionHistoryEntryExtractor.java:
```diff
@@ -0,0 +1,105 @@
+
+        Preconditions.checkArgument(
+                entry != null,
+                "The passed ExecutionVertexId {} does not correspond to a failed execution.",
```
this can be misleading; e.g., if executionJobVertices does not contain the vertex.

XComp replied:
This is fixed by adding explicit failures 👍

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/ExceptionHistoryEntryExtractor.java:
```diff
@@ -0,0 +1,105 @@
+                executionJobVertices.get(executionVertexID.getJobVertexId());
+
+        if (executionJobVertex == null) {
+            return null;
```
Why isn't this failing immediately?

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/ExceptionHistoryEntryExtractor.java:
```diff
@@ -0,0 +1,105 @@
+        final ExecutionVertex[] executionVertices = executionJobVertex.getTaskVertices();
+
+        if (executionVertices.length <= executionVertexID.getSubtaskIndex()) {
+            return null;
```
Why isn't this failing immediately?

XComp commented:
I addressed all comments and marked the PR as reviewable. The AzureCI failure seems to be unrelated [FLINK-21929](https://issues.apache.org/jira/browse/FLINK-21929)).

XComp commented:
I added a separate commit to cover adding the timestamp to the `FailureHandlingResult`. I kept the commit separately to make it easier to review the two issues (the actual purpose of this PR and the timestamp refactoring).

I verified manually that the UI is not affected by this change. @zentol please give it another round...

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/ExceptionHistoryEntry.java:
```diff
@@ -40,45 +39,20 @@
-        return new ExceptionHistoryEntry(
-                failureInfo.getException(),
-                failureInfo.getTimestamp(),
+    @VisibleForTesting
```
this is incorrect

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/ExceptionHistoryEntryExtractor.java:
```diff
@@ -0,0 +1,149 @@
+            final ExecutionVertex executionVertex =
+                    getExecutionVertex(executionJobVertices, otherExecutionVertexId);
+            if (executionVertex.getFailureInfo().isPresent()) {
+                root.add(
```
why are we mutating the entry instead of first collecting all concurrent failures?

XComp replied:
Right now, we're benefiting from having a common interface for instantiating the `RootExceptionHistoryEntry` and the `ExceptionHistoryEntry` in the [ExceptionHistoryEntryExtractor:127](https://github.com/apache/flink/blob/5b15b5614d7eacc7a2bb4cabfa6a12391921f9fd/flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/ExceptionHistoryEntryExtractor.java#L127). Hence, I went ahead with your other proposal of reducing the visibility of the `add` method.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/RootExceptionHistoryEntry.java:
```diff
@@ -0,0 +1,68 @@
+     * @return The {@code ExceptionHistoryEntry} instance.
+     */
+    @VisibleForTesting
+    public static RootExceptionHistoryEntry fromGlobalFailure(Throwable cause, long timestamp) {
```
it was surprising that there is no `fromLocalFailure` flavor, that would also enforce the task name / location to not be null.

XComp replied:
Good idea. Thanks for that. 👍

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/RootExceptionHistoryEntry.java:
```diff
@@ -0,0 +1,68 @@
+        super(cause, timestamp, failingTaskName, taskManagerLocation);
+    }
+
+    public void add(ExceptionHistoryEntry concurrentException) {
```
```suggestion
    void add(ExceptionHistoryEntry concurrentException) {
```
Ideally we get rid of it though.

zentol reviewed (commented):
There are some CI failures that should be looked at.

XComp commented:
The `JobMasterTest` failed due to a `NullPointerException` caused by me adding a notNull check to the `FailureHandlingResult` as part of my initial efforts to introduce `ErrorInfo` in `FailureHandlingResult` instead of dedicated members for the `failureCause` and `timestamp`. I reverted that change due to problems with the `SerializedThrowable` swallowing annotations that is necessary for [identifying non-recoverable errors](https://github.com/apache/flink/blob/c77a686c195d1742c276f4a9e75899c8b85377bb/flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandler.java#L140). I reverted the null check.

XComp commented:
Might writing down the previous comment got me thinking again: Ideally, we would want to have this null check since a failure should always have a cause. We didn't introduce a null check so far because of [FLINK-21376](https://issues.apache.org/jira/browse/FLINK-21376). There is [ErrorInfo. createErrorInfoWithNullableCause](https://github.com/apache/flink/blob/c77a686c195d1742c276f4a9e75899c8b85377bb/flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ErrorInfo.java#L50) for handling this. But it feels to be handled in the wrong place. Instead, we should substitute the `null` value by the `FlinkException` in [TaskExecutionState's constructor](https://github.com/apache/flink/blob/7f1853d3258c38b24e4be36adcfe08a59f211f39/flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/TaskExecutionState.java#L68). This way, we make sure that the `null` never ends up in the `ExecutionGraph`-related code which enables us to make the invariants stricter on our end.

zentol commented:
But that's all part of FLINK-21376, correct? FYI, I don't think the TaskExecutionState should set an exception on it's own, instead it should enforce that if the exception variants are used the exception is not null.

zentol reviewed (approved):
Would you consider that blocking the PR, or you can we go ahead with the merge?

XComp commented:
> But that's all part of [FLINK-21376](https://issues.apache.org/jira/browse/FLINK-21376), correct? FYI, I don't think the TaskExecutionState should set an exception on it's own, instead it should enforce that if the exception variants are used the exception is not null.

It would be some preparation/cleanup task for [FLINK-21376](https://issues.apache.org/jira/browse/FLINK-21376). I created [FLINK-22060](https://issues.apache.org/jira/browse/FLINK-22060) to cover this. Hence, it's not blocking this PR.
